alacritty_terminal/src/selection.rs

```diff
@@ -283,18 +176,55 @@ impl Selection {
                 start.col -= 1;
             }
         }
-
-        // Remove first cell if selection starts at the right of a cell
-        if end_side == Side::Right && start != end {
+        // Remove last cell if selection ends at the right of a cell
+        if end_side == Side::Right && start != end && end_type == SelectionType::Simple {
             end.col += 1;
         }
 
-        if alt_screen {
-            Selection::alt_screen_clamp(&mut start, &mut end, lines, cols)?;
+        // Semantic expansion
+        if start_type == SelectionType::Semantic {
+            start = Point::from(term.semantic_search_right(start.into()));
+        }
+        if end_type == SelectionType::Semantic {
+            end = Point::from(term.semantic_search_left(end.into()));
+        }
+
+        // Line expansion
+        if start_type == SelectionType::Lines {
+            start = Point { line: start.line, col: cols - 1 };
+        }
+        if end_type == SelectionType::Lines {
+            end = Point { line: end.line, col: Column(0) };
         }
 
-        // Return the selection with all cells inclusive
-        Some(Span { start: start.into(), end: end.into() })
+        let span = Some(Span { start: start.into(), end: end.into() });
+
+        // Expand selection across double-width cells
+        span.map(|mut span| {
+            let grid = term.grid();
+
+            if span.end.col < grid.num_cols()
+                && grid[span.end.line][span.end.col].flags.contains(Flags::WIDE_CHAR_SPACER)
+            {
+                span.end.col = Column(span.end.col.saturating_sub(1));
+            }
+
+            if span.start.col.0 < grid.num_cols().saturating_sub(1)
+                && grid[span.start.line][span.start.col].flags.contains(Flags::WIDE_CHAR)
+            {
+                span.start.col += 1;
+            }
+
+            span
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        if self.start_type == SelectionType::Simple && self.end_type == SelectionType::Simple {
+            self.region.start == self.region.end
+        } else {
+            false
+        }
```

Comment:
This body can be changed to `self.start_type == SelectionType::Simple && self.end_type == SelectionType::Simple && self.region.start == self.region.end`.